src/transformers/models/dpt/modeling_dpt.py

```diff
@@ -0,0 +1,1108 @@
+            attentions=outputs.attentions,
+        )
+
+
+class DPTSemanticSegmentationHead(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+
+        self.config = config
+
+        features = config.fusion_hidden_size
+        self.head = nn.Sequential(
+            nn.Conv2d(features, features, kernel_size=3, padding=1, bias=False),
+            nn.BatchNorm2d(features),
+            ACT2FN["relu"],
+            nn.Dropout(config.semantic_classifier_dropout),
+            nn.Conv2d(features, config.num_labels, kernel_size=1),
+            nn.Upsample(scale_factor=2, mode="bilinear", align_corners=True),
+        )
+
+    def forward(self, hidden_states: List[torch.Tensor]) -> torch.Tensor:
+        # use last features
+        hidden_states = hidden_states[self.config.in_index]
+
+        logits = self.head(hidden_states)
+
+        return logits
+
+
+class DPTAuxiliaryHead(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+
+        features = config.fusion_hidden_size
+        self.head = nn.Sequential(
+            nn.Conv2d(features, features, kernel_size=3, padding=1, bias=False),
+            nn.BatchNorm2d(features),
+            ACT2FN["relu"],
+            nn.Dropout(0.1, False),
+            nn.Conv2d(features, config.num_labels, kernel_size=1),
+        )
+
+    def forward(self, hidden_states):
+        logits = self.head(hidden_states)
+
+        return logits
+
+
+@add_start_docstrings(
+    """
+    DPT Model with a semantic segmentation head on top e.g. for ADE20k, CityScapes.
+    """,
+    DPT_START_DOCSTRING,
+)
+class DPTForSemanticSegmentation(DPTPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.dpt = DPTModel(config, add_pooling_layer=False)
+
+        # Neck
+        self.neck = DPTNeck(config)
+
+        # Segmentation head(s)
+        self.head = DPTSemanticSegmentationHead(config)
+        self.auxiliary_head = DPTAuxiliaryHead(config) if config.use_auxiliary_head else None
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def compute_loss(self, logits, auxiliary_logits, labels):
+        # upsample logits to the images' original size
+        upsampled_logits = nn.functional.interpolate(
+            logits, size=labels.shape[-2:], mode="bilinear", align_corners=False
+        )
+        if auxiliary_logits is not None:
+            upsampled_auxiliary_logits = nn.functional.interpolate(
+                auxiliary_logits, size=labels.shape[-2:], mode="bilinear", align_corners=False
+            )
+        # compute weighted loss
+        loss_fct = CrossEntropyLoss(ignore_index=self.config.semantic_loss_ignore_index)
+        main_loss = loss_fct(upsampled_logits, labels)
+        auxiliary_loss = loss_fct(upsampled_auxiliary_logits, labels)
+        loss = main_loss + self.config.auxiliary_loss_weight * auxiliary_loss
+
+        return loss
```

Comment:
We had the debate on another model recently, and IMO this should all be in the forward pass.

Reply:
Imo, it's way cleaner to not include this in the forward pass. The same is done in BEiT. (as this one also uses an auxiliary head).

cc @FrancescoSaverioZuppichini

Reply:
I don't find this cleaner, it makes the reader go up and down in the modeling file for 7 lines of code while introducing inconsistency with every other model in the library.

BEiT should also not have defined this separate `compute_loss` function IMO.

Reply:
I also find this is cleaner and I'd personally prefer to read a file defined in this way (but this is personally). I don't have the experience to tell if it is negligible to have this kind of inconsistency thou.

Reply:
+1 for Sylvain from my side. I really think it's important to also keep the "Principal of least astonishment": https://en.wikipedia.org/wiki/Principle_of_least_astonishment into consideration here a bit as I've never seen any other model that has a `compute_loss` function. Changing the design here comes with a big cost of "astonishment" IMO for everybody that is used to how the loss is computed previously / in other models

Reply:
Ok I have removed the `compute_loss` method, it's now all included in the forward pass.